### DISCUSSION TITLE
(How) can one implement a `rich.Prompt` using a `textual.widgets.Input`?

### DISCUSSION THREAD

**Discussion by mrossinek** [Textualize/textual#1784] [Q&A] [answered]
I am writing an application which provides both, a CLI (using `rich`) and a TUI (using `textual`). This application provides various commands, some of which can require user interaction while they are executed. This brings me to the use case that I would like to be able to replace a `rich.Prompt` object (which gets triggered during the CLI) with a `textual.widgets.Input` workflow when this command is triggered from the TUI. However, I am having a hard time realizing this.

I have tried to create a minimal code example to show how far I have gotten and attached it below:
```python
import argparse

from rich.prompt import Confirm, PromptBase
from textual.app import App
from textual.widgets import Input, Static


class TextualPrompt(PromptBase):
    tui = None

    @classmethod
    def get_input(cls, console, prompt, password, stream=None):
        inp = Input(value="test")
        cls.tui.mount(inp)
        inp.focus()
        # TODO: how can I await the Input.Submitted event here to return its result?


class AskCommand:
    def __init__(self, prompt=Confirm):
        self.prompt = prompt

    def execute(self):
        return self.prompt.ask("Do you love Textualize?")


class PromptApp(App):
    BINDINGS = [("a", "ask", "Ask")]

    def compose(self):
        yield Static("Hello from textual!")

    def on_input_submitted(self, event):
        """TODO."""
        event.sender.remove()
        return event.value

    async def action_ask(self):
        TextualPrompt.tui = self
        ask = AskCommand(TextualPrompt)
        res = ask.execute()
        if res:
            await self.mount(Static("Awesome! Make sure to star our Github repo!"))
        else:
            await self.mount(Static("Let us know how we can improve!"))


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("interface", choices=["cli", "tui"])

    args = parser.parse_args()

    if args.interface == "cli":
        ask = AskCommand()
        print(ask.execute())
    elif args.interface == "tui":
        PromptApp().run()
```

The `TODO` marker in `TextualPrompt.get_input` is what is causing me the most headaches.
On top of that, I wonder how/if I could use `await` within that method, but this requires the method to be declared as `async` which leads to a `RuntimeWarning` because the `get_input` method is never awaited. I am not quite sure how I am supposed to handle this because I would need to mark every method as `async` if I propagate the `await` down into the call of `ask.execute()`...

Any pointers/ideas/help is welcome :slightly_smiling_face:

**Comment by willmcgugan** [comment]
I think what you would need to do is have an app that accepts the input, then call `self.exit(result)` in the `on_input_submitted` method.

That way you can `return MyApp().run()` from your prompt method.

**Reply by mrossinek** [reply]
Hm I see :thinking: Maybe my example had been reduced a bit too much because I actually want the app to do more than just prompt i.e. it should persist after this particular command finished.

Is there a way to nest apps? If so, this might work. But even then I would need to wait for the app to exit before the `get_input` method can return, right?

<hr/>

I did also try playing around with making `TextualPrompt` inherit from both `PromptBase` and `Input` because this would allow me to implement `on_input_submitted` on that class directly. But this is proving rather difficult as one might imagine..

**Comment by mrossinek** [comment, marked as answer]
After quite some tinkering I think I have found a way to make this work. Looks like there are multiple key aspects:
1. I need to create a separate asyncio task for my binding action to make sure the main App event loop does not get blocked
2. I can unmount my `Input` widget upon `on_input_submitted`
3. I can use `App.is_mounted(widget)` to check whether or not my input object has been unmounted during my `get_input` method

I am posting the code below in case you have some suggestions on how it might be improved further:
```python
import argparse
import asyncio

from rich.prompt import Confirm, InvalidResponse
from textual.app import App
from textual.widgets import Input, Static


class TextualPrompt(Confirm):
    tui = None

    async def __call__(self, *, default=..., stream=None):
        while True:
            self.pre_prompt()
            prompt = self.make_prompt(default)
            value = await self.get_input(self.console, prompt, self.password, stream=stream)
            if value == "" and default != ...:
                return default
            try:
                return_value = self.process_response(value)
            except InvalidResponse as error:
                self.on_validate_error(value, error)
                continue
            else:
                return return_value

    @classmethod
    async def get_input(cls, console, prompt, password, stream=None):
        inp = Input()
        await cls.tui.mount(inp)

        inp.focus()

        while cls.tui.is_mounted(inp):
            await asyncio.sleep(0)

        return inp.value


class AskCommand:
    def __init__(self, prompt=Confirm):
        self.prompt = prompt

    def execute(self):
        return self.prompt.ask("Do you love Textualize?")


class PromptApp(App):
    BINDINGS = [("a", "ask", "Ask")]

    def compose(self):
        yield Static("Hello from textual!")

    def on_input_submitted(self, event):
        event.sender.remove()

    def action_ask(self):
        asyncio.create_task(self._async_action_ask())

    async def _async_action_ask(self):
        TextualPrompt.tui = self
        ask = AskCommand(TextualPrompt)
        res = await ask.execute()
        if res:
            await self.mount(Static("Awesome! Make sure to star our Github repo!"))
        else:
            await self.mount(Static("Let us know how we can improve!"))
        self.refresh()


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("interface", choices=["cli", "tui"])

    args = parser.parse_args()

    if args.interface == "cli":
        ask = AskCommand()
        res = ask.execute()
        if res:
            print("Awesome! Make sure to star our Github repo!")
        else:
            print("Let us know how we can improve!")
    elif args.interface == "tui":
        PromptApp().run()
```
